Update unit tests information with examples

eLeontev opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ * ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ * ] Bugfix
- [ ] Feature
- [ * ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ * ] Other... Please describe: add section related to unit-tests with pure javascript approach


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Existed doc describes complex approaches to write unit tests including testBed (moslty for integration and e2e) and does not contain any information how to write unit-tests a whole matched to angular app.

Issue Number: N/A


## What is the new behavior?
add examples and description of main approaches how to write unit-tests in vanila javascript using angular components

## Does this PR introduce a breaking change?

- [ ] Yes
- [ * ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

eLeontev committed:
docs: fix styles issue, fix const duplication

eLeontev committed:
docs: add unit-tests for angular component as pure javaScript instance with comments

eLeontev committed:
docs: add section with explanation how to cover angular application as pure javascript app with examples

eLeontev commented:
@IgorMinar hi, I'm not tech-writer, so the description part can contain misprints and not be clear enough. Please put the comments with the format which you prefer to improve it.

eLeontev committed:
docs: add link in the docs to written test

eLeontev committed:
docs: fix validation error

eLeontev commented:
@mhevery @IgorMinar hi, could you clarify will this PR be reviewed and when?

kapunahelewong commented:
Thank you, @eLeontev for all of this work! It must have taken some time to do. Though this is a docs related PR, we'll need engineering review first. After the appropriate reviewer(s) have given approval, then the docs team can review to help you get it inline with the docs style guide that we use. I'll mark it for review so that it's in the queue for engineering. 

So you know, ultimately we try to make sure that most of the code snippets that we use come from the example files (this is a work in progress!🙂). If you're not sure how to accomplish that, when we do the docs review I can help you with it, but that will be something to keep in mind for a future commit.

If you have any questions along the way, feel free to tag me. 🌻

petebacondarwin reviewed (commented):
Hi @eLeontev - thanks for putting this PR together. I am sorry it has been waiting around for a long time.
I think it is now somewhat outdated, since the testing guides got considerably re-worked over the summer.
As such, I wonder if it is still relevant. There are certainly areas in the guide that talk about basic unit testing.

If this PR is going to land there a lot of small improvements that i think would need to be made, but let's hear from @aikidave about the overall idea before we dig too much into the detail.

aikidave commented:
@TeriGlover, could you please review the testing.md file in this pull request? Thanks!

eLeontev commented on aio/content/examples/testing/src/app/hero/hero-list.component.no-testbed.spec.ts:
```diff
@@ -0,0 +1,88 @@
+import { HeroListComponent } from './hero-list.component';
+
+describe('HeroListComponent', () => {
+  let component: any;
+
+  let router: any;
+  let heroService: any;
```
ok, my point is about with TS we have strong types flow, it's the benefit for production. In other hand using types in tests requires us to follow the same data structures even if they are extremely huge.

So it's the trade off we need to solve
with using types:
1. compilation of the tests will fail one method related types are changed  (we can easy define in IDE the issue with the type in the test and fix it)
2. using the same data structures (e.g. extra complex json, etc.) as in the tested method are required

without types:
1. tests will fail during run (e.g. CI). Lack of types will not highlight failed tests in IDE before run the test 
2. we can use any types only required to tested method to simplify our test as far as it can be simplified.

Based on my experience usage of types is good when you work with small data structures, once the app will be huge enough it will require `MOCK_DATA.entity.json/ts` files. But more interesting point is it will lead performance issues in the tests. (e.g. for custom type we can work with primitives and use `toBe` equal by link - perfromance is good, but for FP approach or if we simply prefer to work in immutable approach we will use `toEqual` or `toStrickEqual` which affect the test performance a lot due to comparation based on value).

Example with complexity: I was on the project where was strong following one rule (max-lines-per-file: 200) and with using types in the test we were necessary to split our tests to test ~each method per file to face the linter criteria (of course it's discussion question, but point is that the reason of creation multiple files for test was extremely huge data structure based on business complexity )

I've just shared my point and pluses/cons both approaches.

petebacondarwin replied:
Consider the `router` variable. By using `any`, and then initializing it to `{}`, you are making assumptions about what the `HeroListComponent` might choose to do with the value when you pass it to the constructor, making the test brittle in case a later developer changes the code in the constructor.

If you really know this is going to be some opaque object, then I would consider typing it as `unknown`, which would mean that no one can accidentally try to use this variable as a real `Router` without explicitly casting it. Another option is to define the `router` variable with a proper type (e.g. `Router`) but then when assigning the dummy value, you cast at that point:

```
let router: Router;

beforeEach(() => {
  router = {} as any;
});
```

Then, at least, any usage of the `router` variable is checked.

While `any` can provide shortcuts and possible performance benefits, the loss of typing support outweighs it most cases.

eLeontev replied:
no `any` for now since `js` file is used

eLeontev commented on aio/content/examples/testing/src/app/hero/hero-list.component.no-testbed.spec.ts:
```diff
@@ -0,0 +1,88 @@
+    component = new HeroListComponent(router, heroService);
+  });
+
+  // optional to define prams defined in constructor
+  it('should define properties on instance initialization', () => {
+    expect(component.hasSpecificProperyOrMethod).toBeUndefined();
+  });
```
I did not want to change example-production file, so set example validation how it can be and had guessed it would be clear. Let me share it in the comment:
```
class TestedClass {
  private readonly debounceDelayMs= 300;
  private flagWithFalseByDefault = false;
}

// in tests
let testedInstance: any;

beforeEach(() => {
  testedInstance = new TestedClass();
})

it('should definedebounce delay on instance initialization', () => {
  expect(testedInstance.debounceDelayMs).toBe(300)
});

it('should init flag with false by default', () => {
   expect(testedInstance.flagWithFalseByDefault ).toBefalsy()
})
```

eLeontev replied:
about compile - I still use any, so no compilation errors

eLeontev replied:
move this test to external file with examples and prepare proper test data for this scenario

eLeontev commented on aio/content/examples/testing/src/app/hero/hero-list.component.no-testbed.spec.ts:
```diff
@@ -0,0 +1,88 @@
+    // define dependency's behavior
+    beforeEach(() => {
+      // use primitives to simplify comparison with expected results
+      heroService.getHeroes.and.returnValue(heroes);
```
can you clarify the difference for `#getHeroes` between original  `Heroes` type/structure and simple mocked?

For me it has no difference since the tested method does not communicate with returned type and for it it does not matter which type /data it returns, more other in case if data strcuture/type will be changed in the app, but `getHeroes` will not be updated (e.g. it still will match to business criteria) - we do not need to fix this test since it still works as expected.

eLeontev replied:
yep, it's the point to highlight that if production code is changed the test must highlight that it need to be also actualized on the test side to make sure that the changes done in the code are valid.

eLeontev commented:
in case if this PR have a chance to be merged, please ping me and I will rebase it.

TeriGlover commented:
Hi Dave,
I don't see a Mary Poppins link (probably not on this test) but It's
unclear where the text is that I should edit.

On Tue, Jan 26, 2021 at 7:05 AM eLeontev <notifications@github.com> wrote:

> in case if this PR has a chance to be merged, please ping me and I will
> rebase it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/36202#issuecomment-767604573>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AR4DPGX5ZWX2BFFHKS5MFGTS33K3BANCNFSM4LR6VMSQ>
> .
>


-- 
*--Teri Glover--*
*Technical Editor*

mary-poppins commented:
You can preview 1dce77c at https://pr36202-1dce77c.ngbuilds.io/.
You can preview 87c808b at https://pr36202-87c808b.ngbuilds.io/.

gkalpak commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+
+ spy(anotherParam);
+ expect(spy).toHaveBeenCalledWith(param); // failed #valid
+```
````
Having hard-coded snippets (such as these ones here) is something we are trying to avoid. It makes it easy to introduce typos and makes the content easier to break in the future without us noticing.

Ideally, all code examples should come from doc-regions (which are part of a tested mini-app).

aikidave replied:
I agree with @gkalpak. We should not have hard-coded examples in our documentation--even though I know how much easier it is (at least in the short term!) to do so.

In addition, I think this text needs a little more work before we can publish:

* Can we title the section to reflect the problem this content is trying to solve?
* In the opening paragraph, can we provide more clarity about why this content might be relevant to the reader?
* The principles you refer to don't seem to require an ordered list. Should users consider these principles in order?
* Can we add more explanation that clarify these principles? Why are they important?
* What is the difference between common tips and the principles?
* Most importantly, we definitely will need some feedback to ensure that we agree on these principles and common tips.

I'm happy to chat about how to implement some of these changes if you like! I realize this might seem like a lot of feedback, but I don't think it will take long to implement, provided that we get engineering sign-off on the content!

gkalpak reviewed (commented):
Drive-by review 🚗

TeriGlover commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+
+In the test we need to follow the same principe to simplify it.
+
+1. each tested method should be wrapped to the individual scope with initialization on the root level to perform it before each tests
```
```suggestion
1. Each tested method should be wrapped to the individual scope with initialization on the root level to perform it before each test.
```
Clarification needed. Is this saying:
 ...with initialization performed before each test at the root level.   ?

eLeontev replied:
The description is re-written. I also update the example.
![image](https://user-images.githubusercontent.com/15786916/116901900-29436100-ac43-11eb-88b4-38c806f94de9.png)

TeriGlover commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ });
+```
+
+2. Each case should be performed with not dirty state
````
```suggestion
2. Each case should be performed with not dirty state.
```
Is there another name for "not dirty state" ?

eLeontev replied:
I don't know usually it is used as `isDirty = false` by default. I guess it is used in common state as not dirty without special word.
The missed dot is added.

TeriGlover commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ expect(testedMethod(comfortableMockedParams, mockedDataWithDefinedMethod)).toBe(comfortablePredictableTestResult);
+```
+
+4. Each dependency and data used in the tested method should be mocked or replaced to the spy
````
```suggestion
4. Each dependency and data used in the tested method should be mocked or replaced to the spy.
```
Unclear. Please reword.

eLeontev replied:
The missed dot is added.  
In each method we have dependencies and data used inside them so the y must be replaced to mocked values for static data and spies for methods/functions.

TeriGlover commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ });
+ ```
+
+5. Each type should be simplified as possible
````
```suggestion
5. Each type should be simplified as possible.
```
Ambiguous; please clarify. Is this saying: 
Each type should be as simple as possible
Each type should be made as simple as possible
Each type should be simplified as much as possible

eLeontev replied:
updated to `Each type should be simplified as much as possible.` For me all these are close for meaning. Anyway I meant in the test we need only types used inside the tested methods, so all passed but not used in them should be omitted.
```
type A = {stringProperty: string; numberProperty: number}
const getStrignProperty= (a: A) => a.stringProperty;

//in the test
expect(getStrignProperty({stringProperty: 'stringProperty'})).toBe('stringProperty'); // numberProperty is not defined as not used in the tested method
``

TeriGlover commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ const productionMethod1 = (params: ComplicatedDataType, dependency: any) => dependency.handler(params);
+ const productionMethod2 = ({ nestedComplicatedDataTypes }: ComplicatedDataType, {handler}: any) => handler(nestedComplicatedDataTypes);
+ 
+ // use any to avoid following complicated data structure to reduce mock/stub data size and simplify tests
```
```suggestion
 // use any to avoid following complicated data structure to reduce mock/stub data size and simplify tests
```
Sentence needs a noun after "Use any (noun) to avoid following.."

eLeontev replied:
`any` is noun in this case. anyway updated to `// use `any` type to avoid following complicated data structure to reduce mock/stub data size and simplify tests`

TeriGlover commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ expect(handler).toHaveBeenCalledWith(nestedComplicatedDataTypes);
+```
+
+6. Tests should avoid any mutations
````
```suggestion
6. Tests should avoid any mutations.
```

eLeontev replied:
The dot is added.

TeriGlover commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ })
+```
+
+7. Common tips
````
```suggestion
7. Common tips:
```

eLeontev replied:
colon is added

TeriGlover commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+```
+
+7. Common tips
+ - do not forget any angular's component (e.g. services/directives etc.) are simple classes (function constructor) with specific annotations
````
```suggestion
 - Do not forget that any Angular components (such as services, directives, etc.) are simple classes (function constructors) with specific annotations.
```
For clarity.

eLeontev replied:
Updated to proposed text.

TeriGlover commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ let dependencyMethodSpy: any; // invalid
+```
+ 
+ - use named only spies and directly assigned dependencies methods for auto-completing and simpler debug 
````
```suggestion
 - Use named-only spies and directly assigned dependencies methods for auto-completing and simpler debugging. 
```
For clarity.

eLeontev replied:
Updated to proposed text.

TeriGlover reviewed (approved):
Editing review to clarify wording in several places.

TeriGlover commented:
This one PR has had an editing pass to clarify wording in several places;
still needs attention.

On Wed, Feb 3, 2021 at 12:02 PM David Shevitz <notifications@github.com>
wrote:

> *@aikidave* requested changes on this pull request.
> ------------------------------
>
> In aio/content/guide/testing.md
> <https://github.com/angular/angular/pull/36202#discussion_r569709932>:
>
> > + }
> +  // DI.-> auto com-completing
> +```
> +
> + - do not forget validate arguments to prevent regressions in their changes
> +```typescript
> + let spy = jasmine.createSpy('spy');
> + spy(param);
> + expect(spy).toHaveBeenCalled(); // successful #invalid
> +
> + spy(param);
> + expect(spy).toHaveBeenCalledWith(param); // successful #valid
> +
> + spy(anotherParam);
> + expect(spy).toHaveBeenCalledWith(param); // failed #valid
> +```
>
> I agree with @gkalpak <https://github.com/gkalpak>. We should not have
> hard-coded examples in our documentation--even though I know how much
> easier it is (at least in the short term!) to do so.
>
> In addition, I think this text needs a little more work before we can
> publish:
>
>    - Can we title the section to reflect the problem this content is
>    trying to solve?
>    - In the opening paragraph, can we provide more clarity about why this
>    content might be relevant to the reader?
>    - The principles you refer to don't seem to require an ordered list.
>    Should users consider these principles in order?
>    - Can we add more explanation that clarify these principles? Why are
>    they important?
>    - What is the difference between common tips and the principles?
>    - Most importantly, we definitely will need some feedback to ensure
>    that we agree on these principles and common tips.
>
> I'm happy to chat about how to implement some of these changes if you
> like! I realize this might seem like a lot of feedback, but I don't think
> it will take long to implement, provided that we get engineering sign-off
> on the content!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/36202#pullrequestreview-582752160>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AR4DPGWXWNNHSIBHKO7432LS5GTWHANCNFSM4LR6VMSQ>
> .
>


-- 
*--Teri Glover--*
*Technical Editor*

petebacondarwin commented:
@eLeontev - thanks for your patience on this PR. There are various outstanding comments that need to be addressed and the PR needs to be rebased on master to resolve the conflicts. Would you fix up all the textual edits, and then if you need help with creating real examples rather than inline ones, let me know.

Commit:
Merge branch 'master' into update_unit_tests_info

# Conflicts:
#	aio/content/guide/testing.md

eLeontev committed:
docs: fix merge conflicts and move new docs to separate guide file to follow new docs structure. Add link from main test page to moved created guide.

eLeontev committed:
docs: fix typos, improve descriptions and align guide styles based on feedback

eLeontev committed:
docs: revert changes for affected test file

eLeontev committed:
docs: update styles to proposed after fix code errors in the file. Add declaration of `activatedRoute`.

eLeontev commented on aio/content/examples/testing/src/app/hero/hero-detail.component.no-testbed.spec.ts:
```diff
@@ -13,20 +13,24 @@ describe('HeroDetailComponent - no TestBed', () => {
+    };
+    activatedRoute = new ActivatedRouteStub({
+      id: expectedHero.id,
+    });
```
Updated to proposed formatting. Btw I changed this file since there were code errors. If you want I could revert my changes to origin.

eLeontev commented on aio/content/examples/testing/src/app/hero/hero-detail.component.no-testbed.spec.ts:
```diff
@@ -40,18 +44,22 @@ describe('HeroDetailComponent - no TestBed', () => {
+    expect(router.navigate.calls.any()).toBe(
+      false,
+      'router.navigate not called yet'
+    );
```
Updated to proposed formatting.

eLeontev commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+```javascript
+ const pureFunction = (arg) => arg;
+
+ const notPureFunction = () => arg;
````
True. Updated to ` const notPureFunction = () => closuredVariable;`

eLeontev commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
 
+## Unit Testing Angular App As Pure JS Application
+
+Unit testing is the low level testing. It's goal is to tests units. Unit means any simple part of application.
```
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+## Unit Testing Angular App As Pure JS Application
+
+Unit testing is the low level testing. It's goal is to tests units. Unit means any simple part of application.
+It can be module as unit / component as unit or method as unit.
```
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+Unit testing is the low level testing. It's goal is to tests units. Unit means any simple part of application.
+It can be module as unit / component as unit or method as unit.
+
+In JavaScript unit-testing mostly means testing of methods. In this context the goal of this testing is to cover <b>only tested methods functionality</b>.
```
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+
+In JavaScript unit-testing mostly means testing of methods. In this context the goal of this testing is to cover <b>only tested methods functionality</b>.
+
+To simplify the tests and make them more stable and predictable they should be independent.
```
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+
+To simplify the tests and make them more stable and predictable they should be independent.
+
+Independence in JavaScript mostly compatible to Functional Programming e.g pure functions.
```
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+To simplify the tests and make them more stable and predictable they should be independent.
+
+Independence in JavaScript mostly compatible to Functional Programming e.g pure functions.
+Each pure function should return the same result with the same passed arguments and does not have any side effects (e.g. closures)
```
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ const notPureFunction = () => arg;
+```
+
+In the test we need to follow the same principe to simplify it.
````
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ });
+ ```
+
+3. Each test should not test another methods in the tested method
````
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
```diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ // static arguments can be defined once too 
+ const params = 'params';
+ 
+ // definition of "dynamic" values should be performed before each tests to clean up test's state 
```
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+ expect(service.serviceMethod).toBe(angularService.serviceMethod);
+```
+
+ - name dependencies and their methods as in production code to simplify their search/debug
````
Updated to proposed text.

eLeontev commented on aio/content/guide/testing.md:
````diff
@@ -2675,6 +2675,242 @@ Explore the options and combinations on your own.
+  // DI.-> auto com-completing
+```
+
+ - do not forget validate arguments to prevent regressions in their changes
````
Updated to proposed text.

eLeontev committed:
docs: update styles, fix typos after feedback, rewrite description to clarify the meaning

eLeontev commented:
I've answered to mostly of the comments. And missed common once like we need to discuss, etc.

eLeontev commented:
@aikidave to your questions:

`1. Can we title the section to reflect the problem this content is trying to solve?` Sure, if my choose is not valid, please provide any alternative or your own best choose.
`2. In the opening paragraph, can we provide more clarity about why this content might be relevant to the reader?` please clarify what do you mean about `more` it would be great if you highlight topic or direction you want to see in each paragraph.
`3. The principles you refer to don't seem to require an ordered list. Should users consider these principles in order?` No, they don't. I ordered them based on my own priority. Anyway all of them are independent. Do you prefer to replace ordered list to not ordered?
`4. Can we add more explanation that clarify these principles? Why are they important?` I've described their importance in the top of the doc about pure approach and immutability. Do you guess more is necessary? if so, please  direct to the topics where you guess the description can be extended.
`5. What is the difference between common tips and the principles?` there are no big diff. I split them like principles are you need to follow to achieve the result to test your application in pure way. And the tips are the list of recommendations could help you to achieve it with less effort. Both of them could be combined to one long list.
`6. Most importantly, we definitely will need some feedback to ensure that we agree on these principles and common tips.` Ok, I'm still waiting for feedback to continue to improve my MR.

petebacondarwin reviewed (changes requested):
I still do not like to see the loss of type-safety in these examples. But if other Angular team members are OK with it, I will accept that.

All the code snippets in the new `testing-as-pure-js-application.md` file need to be moved to proper example files.

petebacondarwin commented:
Hi @eLeontev - do you still want to work on this PR? If so, do you need help moving the examples to external files?

eLeontev commented:
@petebacondarwin  hi, yes i'm ready, but based on the comments for me it's unclear will this MR (like conceptual approach) be merged finally since here are the comments related to accept/reject this MR.

If somebody will confirm that the concept described in the MR is acceptable I continue to work on it.

Could you clarify that this MR with the docs "How to test angular app without dependencies to angular infrastructure" can be accepted?

petebacondarwin commented:
I think conceptually we are happy with the changes. So if we can move the code snippets out into testable code, we have a strong chance of getting this merged. Moreover, I am happy to champion this from our side, if that helps.

petebacondarwin commented:
@eLeontev - do you have time to move the code snippets to external files?

Commit:
Merge branch 'master' into update_unit_tests_info

eLeontev committed:
docs: refactor test file hero-list.component.pure.spec.js

- remove test of not-existing property which is used as example. The proper example will be used in the external test file.
- change file type from `ts` to `js` as casting to any/unknown can be omitted as redundant.

eLeontev committed:
docs: update documentation for unit tests:

- add unit test criteria
- add sections between the text scopes: role/criteria/rules/example to separate them

eLeontev committed:
docs: prepare demo examples to use them in unit tests docs/external files

eLeontev committed:
docs: add demo examples for testing only unit functionality to reuse in  external test file

eLeontev committed:
docs: add pure function example to the demo external file

eLeontev committed:
docs: move all code snippets from docs to external test file and update docs to use external files for tests and demo examples

eLeontev committed:
docs: align test file to used in the docs

eLeontev commented on aio/content/examples/testing/src/app/hero/hero-detail.component.no-testbed.spec.ts:
```diff
@@ -8,12 +8,13 @@ import { Hero } from '../model/hero';
-    expectedHero = {id: 42, name: 'Bubba' };
-    const activatedRoute = new ActivatedRouteStub({ id: expectedHero.id });
+    expectedHero = { id: 42, name: 'Bubba' };
+    activatedRoute = new ActivatedRouteStub({ id: expectedHero.id });
```
all dynamic DI need to initialized  before each test run to avoid test dependencies

example:
```
class Component {
  constructor(private service) {}

  method(shouldSkipCall) {
    if (!shouldSkipCall) {
      this.service.method();
    }
  }
}

describe('Component', () => {
  let component;
  let component2;

  let service;
  const service2 = { method: jasmine.createSpy('method') };

  beforeEach(() => {
    service = { method: jasmine.createSpy('method') };
  });

  beforeEach(() => {
    component = new Component(service);
    component2 = new Component(service2);
  });

  describe('#method', () => {
    it('should call service method if call should not be skipped', () => {
      component.method();
      expect(service.method).toBeCalled();

      component2.method();
      expect(service2.method).toBeCalled();
    });

    it('should not call service method if call should be skipped', () => {
      component.method(true);
      expect(service.method).not.toBeCalled(); // success

      component2.method(true);
      expect(service2.method).toBeCalled(); // failure
    });
  });
});
```

petebacondarwin replied:
The only reason for moving a variable declaration outside the block of the `beforeEach()` is if it is referenced in another block (e.g. in an `it()` block).

But this is not the case for `activatedRoute` it is just a local variable. It does not need to be declared outside this block.

eLeontev commented:
@petebacondarwin  I've done with refactoring. Unfortunately part of your comments are hidden now as attached to no more existing code.

I answered to one which is rest. Please put extra comments again if they are still actual to discuss the proper solution for this PR.

**Updates:**
- add section with unit testing criteria
- update a bit documentation to add section again (to separate the scopes: goals/criteria/example/etc)
- rename `hero.component.pure.spec.ts` to `hero.component.pure.spec.js` - to remove types at all (as I mentioned in the comment above)
- move all functinonal examples to `demo.ts` file to reuse in the docs
- move all test examples to`demo.pure.ts` file to reuse in the docs
- integrated updated examples with docs instead of code-snippets
- add couple of alert comments to the docs with helpful notes

petebacondarwin reviewed (commented):
Hi @eLeontev - thank you for the extra work you have done to move the snippets out into external example code. I am stlll not sure about the loss of typing in tests. I don't think it is a good practice that we should be promoting. But I have asked others on the team to take a view.

petebacondarwin commented:
Hi @eLeontev - thank you for your patience and work that you have put into this PR. I am sorry it has taken so long to do the review and processing of it.

I spoke with other members of the team and we have a number of concerns about this change.

In particular, avoiding the benefit of early failure due to type errors is not a practice we want to encourage. TypeScript is one of our testing tools that gives high quality and early feedback to a developer if something has changed in the code, or if the test is actually passing or receiving incorrect types.

Moreover, the tips in this PR are mostly very generic to any testing situation. We feel that the docs should focus on Angular specific area and leave general testing strategies to be covered by third party content.

So I am afraid that we have decided not to land this change to our docs at this time. Thank you again for the effort you have put into this.
